result.go

```diff
@@ -283,3 +251,96 @@ func (ro resultObject) Extract(rr resultReceiver, v reflect.Value) {
 		f.Result.Extract(rr, v.Field(f.FieldIndex))
 	}
 }
+
+// resultObjectField is a single field inside a dig.Out struct.
+type resultObjectField struct {
+	// Name of the field in the struct.
+	FieldName string
+
+	// Index of the field in the struct.
+	//
+	// We need to track this separately because not all fields of the struct
+	// map to results.
+	FieldIndex int
+
+	// Result produced by this field.
+	Result result
+}
+
+// newResultObjectField(i, f) builds a resultObjectField from the field f at
+// index i.
+func newResultObjectField(idx int, f reflect.StructField) (resultObjectField, error) {
+	rof := resultObjectField{
+		FieldName:  f.Name,
+		FieldIndex: idx,
+	}
+
+	var r result
+	switch {
+	case f.PkgPath != "":
+		return rof, fmt.Errorf(
+			"unexported fields not allowed in dig.Out, did you mean to export %q (%v)?", f.Name, f.Type)
+
+	case isError(f.Type):
+		return rof, fmt.Errorf(
+			"cannot return errors from dig.Out, return it from the constructor instead: "+
+				"field %q (%v) is an error field",
+			f.Name, f.Type)
+
+	case f.Tag.Get(_groupTag) != "":
+		var err error
+		r, err = newResultGrouped(f)
+		if err != nil {
+			return rof, err
+		}
+
+	default:
+		var err error
+		r, err = newResult(f.Type)
+		if err != nil {
+			return rof, err
+		}
+	}
+
+	if rs, ok := r.(resultSingle); ok {
+		// Field tags apply only if the result is "simple"
+		rs.Name = f.Tag.Get(_nameTag)
+		r = rs
+	}
+
+	rof.Result = r
+	return rof, nil
+}
+
+// resultGrouped is a value produced by a constructor that is part of a result
+// group.
+//
+// These will be produced as fields of a dig.Out struct.
+type resultGrouped struct {
+	// Name of the group as specified in the `group:".."` tag.
+	Group string
+
+	// Type of value produced.
+	Type reflect.Type
+}
+
+// newResultGrouped(f) builds a new resultGrouped from the provided field.
+func newResultGrouped(f reflect.StructField) (resultGrouped, error) {
+	rg := resultGrouped{Group: f.Tag.Get(_groupTag), Type: f.Type}
+
+	name := f.Tag.Get(_nameTag)
+	optional, _ := isFieldOptional(f)
+	switch {
+	case name != "":
+		return rg, fmt.Errorf(
+			"cannot use named values with value groups: name:%q provided with group:%q", name, rg.Group)
```

Comment:
In this case, I think "provided" makes sense - result objects come from providers.